### PR TITLE
PREVIEW: supporting a first server push API

### PR DESCRIPTION
**Pull request by jrudolph**
As most of the pieces now came together, I took a first stab at providing a simple low-level server push API.

The idea is that a response can contain the synthetic `Http2PushRequestsHeader` which contains a set of requests to push. These requests are then rendered as PushPromise frames and also reinjected into the normal user-supplied request / response pipeline.

The data flow so far is:

 * original request comes in
 * user-handler produces response including Http2PushRequestsHeader
 * ResponseRendering renders response to new `Http2ResponseSubStream` which includes the rendered kvPairs for pushed requests as well
 * In Demux / Multiplexer: original response in rendered with HEADERS
 * In Demux / Multiplexer: all pushed requests are rendered with PUSH_PROMISE
 * In Demux / Multiplexer: at the same time, all those requests kvPairs are again reinjected as new Http2Substreams
 * RequestParsing will parse those kvPairs again
 * HttpRequests will be dispatched to normal user handler
 * (pushed responses go their normal way)

So, this path currently contains an extra round-trip from HttpRequest -> kvPair -> HttpResponse which we'd probably like to avoid. This can probably most easily done by removing the `httpLayer` stage (which only `map`s its inputs) and just do the mapping directly in the multiplexer.

DO NOT MERGE YET

### CODE REVIEW

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameEvent.scala [hunk 1]
```diff
@@ -75,3 +75,9 @@ final case class ParsedHeadersFrame(
   keyValuePairs: Seq[(String, String)],
   priorityInfo:  Option[PriorityFrame]
 ) extends FrameEvent
+
+final case class ParsedPushPromiseFrame(
+  streamId:         Int,
+  promisedStreamId: Int,
+  keyValuePairs:    Seq[(String, String)]
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/FrameEvent.scala]: `immutable.Seq`?

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala [hunk 2]
```diff
@@ -27,6 +27,10 @@ private[http2] final case class Http2SubStream(initialHeaders: ParsedHeadersFram
   def streamId: Int = initialHeaders.streamId
 }
 
+private[http2] final case class Http2ResponseSubStream(initialHeaders: ParsedHeadersFrame, data: Source[ByteString, Any], pushedRequestsHeaders: Seq[Seq[(String, String)]] = Nil) {
+  def streamId: Int = initialHeaders.streamId
+}
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala]: 👍

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala [hunk 3]
```diff
@@ -61,7 +65,7 @@ object Http2Blueprint {
    * Creates substreams for every stream and manages stream state machines
    * and handles priorization (TODO: later)
    */
-  def demux(): BidiFlow[Http2SubStream, FrameEvent, FrameEvent, Http2SubStream, NotUsed] =
+  def demux(): BidiFlow[Http2ResponseSubStream, FrameEvent, FrameEvent, Http2SubStream, NotUsed] =
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala]: So we'll rename Http2SubStream to `Http2RequestSubStream` I guess

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala [hunk 4]
```diff
@@ -238,11 +238,26 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
       val bufferedSubStreamOutput = new BufferedOutlet[Http2SubStream](substreamOut)
       def dispatchSubstream(sub: Http2SubStream): Unit = bufferedSubStreamOutput.push(sub)
 
+      var _nextPushedStreamId = 2
+      def getNextPushedStreamId(): Int = {
+        val res = _nextPushedStreamId
+        require(res > 0, "Pushed stream IDs overflowed") // FIXME: fail more gracefully
```
Comment (ktoso) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala]: while doing that move to HttpCompliance (and not sure require)

**Comment by ktoso** [comment]
Reviewed the data flow description, it sounds good. Merging the httpLayer into the demux sounds good as well to avoid the additional roundtrips.

File: akka-http2-support/src/test/scala/akka/http/scaladsl/Http2ServerTest.scala [hunk 5]
```diff
@@ -80,33 +95,46 @@ object Http2ServerTest extends App {
...
+         |  <body>
+         |    <h1>Say hello to <i>akka-http-core</i>!</h1>
+         |    <p>Defined resources:</p>
+         |    <ul>
+         |      <li><a href="/ping">/ping</a></li>
+         |      <li><a href="/image-page">/image-page</a></li>
+         |      <li><a href="/crash">/crash</a></li>
+         |    </ul>
+         |  </body>
+         |</html>""".stripMargin), headers = Http2PushRequestsHeader(HttpRequest(uri = "https://localhost:9001/main.css")) :: Nil)
```
Comment (ktoso) [inline akka-http2-support/src/test/scala/akka/http/scaladsl/Http2ServerTest.scala]: cool 👍

File: akka-http2-support/src/main/scala/akka/http/impl/engine/http2/RequestRendering.scala [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+    // scheme, method, path, authority
+    headerPairs += ":method" → request.method.value
+    headerPairs += ":scheme" → request.uri.scheme
+    headerPairs += ":authority" → request.uri.authority.toString.drop(2) // Authority.toString renders leading double slashes see #784
+    headerPairs += ":path" → request.uri.path.toString
+
+    headerPairs ++=
+      request.headers.collect {
+        case header: HttpHeader if header.renderInResponses ⇒ header.lowercaseName → header.value
```
Comment (jrudolph) [inline akka-http2-support/src/main/scala/akka/http/impl/engine/http2/RequestRendering.scala]: Should be `header.renderInRequests`.

**Comment by ktoso** [comment]
TODO: Look at existing servers or proxies what header they use for that.

**Comment by raboof** [comment]
Let's close this until we get back to implementing this. Added a link from #1051 to this PR

**Comment by jrudolph** [comment]
Yep, no reason to keep it open for now.

**Pull request closed without merging**